docs/imds_retry_based_on_errors.md

```diff
@@ -0,0 +1,65 @@
+# MSAL & MSI IMDS Error Handling and Retry Strategy Specification
+
+## Overview
+This document defines the error handling and retry strategy for MSAL when interacting with the IMDS (Instance Metadata Service) endpoint for Managed Identity (MSI) token acquisition.
+
+---
+
+## 1️⃣ HTTP Status Codes & Recommended Actions
+
+| **HTTP Status Code** | **Error Reason**                               | **Recommended Action**                      | **Retry Delay Strategy**                 |
+|----------------------|-----------------------------------------------|---------------------------------------------|-----------------------------------------|
+| **400**             | Bad Request (Invalid Parameters)               | **Do not retry**, fix request               | **No retry**                             |
+| **401**             | Unauthorized                                  | **Do not retry**, check authentication setup | **No retry**                             |
+| **403**             | Forbidden                                     | **Do not retry**, verify permissions       | **No retry**                             |
+| **404**             | IMDS endpoint is updating / Identity Not Found | Retry with Exponential Backoff (max 5 retries) | **1s → 2s → 4s → 8s → 16s**            |
+| **408**             | Request Timeout                                | Retry with Exponential Backoff (max 5 retries) | **1s → 2s → 4s → 8s → 16s (max 16s)**            |
+| **410**             | IMDS is undergoing updates                    | Wait up to **70 seconds**, then retry      | **70s (fixed wait)**                    |
+| **429**             | IMDS Throttle limit reached                   | Retry with Exponential Backoff (max 5 retries) | **2s → 4s → 8s → 16s → 32s (max 32s)**           |
+| **504**             | Gateway Timeout                               | Retry with Exponential Backoff (max 5 retries) | **1s → 2s → 4s → 8s → 16s (max 16s)**            |
```

Comment:
I am worried about 16s delay. It's a long time. It's way longer than the current 3s. Wondering if 8s (i.e. 4 retries) is enough?

Reply:
updated

Reply:
based on discussion with Nidhi, 3 retries with 1-2-4 is better than current 3 retries with 1-1-1